Make memory overlap a little less precise so it works with null data ptr

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Restore storage on meta tensors; increase meta coverage
* #54711 Compute Tensor::toString() without reference to backend
* **#54710 Make memory overlap a little less precise so it works with null data ptr**
* #54709 Make the fuser raise NotImplementedError when unknown device is hit
* #54708 Avoid testing device in cdist when called in a "Math" context
* #54707 Make convolution_overrideable default implementation raise NotImplementedError
* #54706 Disambiguate where 'Doesn't run' error message comes from
* #54705 Fix missing format string qualifier
* #54704 Rename linalg.det test so that it generates a valid method name
* #54703 Add empty_generic, which allocates an empty tensor in a device-generic way
* #54597 Fix some more native_functions.yaml mistakes
* #54627 Nasty little hack to preserve NotImplementedError raised in interpreter

I'm going to make meta tensors have storage (but DataPtr is always
null) so that I can accurately report memory overlap error checking, but
I now have a problem which is that if memory overlap test looks at the
actual data pointer, everything is going to look like it aliases!  A
more conservative test is to just see if the Storage objects themselves
alias, and assume that the data pointers are unique if they don't.

The loss of precision stems from if you unsafely have two distinct
storage objects that point to the same data pointer.  This situation
is pretty rare and so I think it is worth it (and I am hoping no tests
trigger by this.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338810](https://our.internmc.facebook.com/intern/diff/D27338810)

ezyang committed:
Make memory overlap a little less precise so it works with null data ptr

I'm going to make meta tensors have storage (but DataPtr is always
null) so that I can accurately report memory overlap error checking, but
I now have a problem which is that if memory overlap test looks at the
actual data pointer, everything is going to look like it aliases!  A
more conservative test is to just see if the Storage objects themselves
alias, and assume that the data pointers are unique if they don't.

The loss of precision stems from if you unsafely have two distinct
storage objects that point to the same data pointer.  This situation
is pretty rare and so I think it is worth it (and I am hoping no tests
trigger by this.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh reviewed (approved):
LGTM. It might be worth adding a comment somewhere, maybe in `StorageImpl.h`, that BAD stuff will officially happen if you create two Storage objects pointing to the same underlying data? It sounds like that was sort of questionable-but-not-necessarily-bad before.

ezyang commented:
> LGTM. It might be worth adding a comment somewhere, maybe in StorageImpl.h, that BAD stuff will officially happen if you create two Storage objects pointing to the same underlying data? It sounds like that was sort of questionable-but-not-necessarily-bad before.

I would have hoped this would have been obvious, haha. There are more things that won't work if you do this:

* An ordinary deleter on such a storage is wrong, because normal deleters assume unique ownership, but if you have two storages at the same data, that implies there is some sort of shared ownership. So your deleter would have to actually be internally doing some sort of refcount thing
* Deepcopy in Python side relies on storage equality and not data pointer equality; so if there are two separate storages pointing to the same data, the data will actually get duplicated in that case (one data ptr before, two data ptrs after)
* Version counts won't work correctly, because we do all VC tracking at the level of storages (unless you explicitly disconnect the VC with detach); mutation because data pointers are the same are totally untracked

ezyang committed:
Update on "Make memory overlap a little less precise so it works with null data ptr"


I'm going to make meta tensors have storage (but DataPtr is always
null) so that I can accurately report memory overlap error checking, but
I now have a problem which is that if memory overlap test looks at the
actual data pointer, everything is going to look like it aliases!  A
more conservative test is to just see if the Storage objects themselves
alias, and assume that the data pointers are unique if they don't.

The loss of precision stems from if you unsafely have two distinct
storage objects that point to the same data pointer.  This situation
is pretty rare and so I think it is worth it (and I am hoping no tests
trigger by this.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27338810](https://our.internmc.facebook.com/intern/diff/D27338810)

[ghstack-poisoned]
